File: Tutorial-1/insertionSort/insertion_sort.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <algorithm>

std::vector<int> create_array(int size) {
    std::vector<int> arr;
    arr.reserve(size);
    for (int i = 0; i < size; ++i) {
        arr.push_back(i * i);
    }
    return arr;
}

void insertion_sort(std::vector<int>& arr) {
    
    //WRITE YOUR CODE HERE

    
}

int main() {
    const int START_SIZE = 10000;
    const int END_SIZE = 100000;
    const int STEP_SIZE = 10000;

    for (int list_size = START_SIZE; list_size <= END_SIZE; list_size += STEP_SIZE) {
        std::vector<int> numbers = create_array(list_size);

        auto start = std::chrono::high_resolution_clock::now();

        insertion_sort(numbers);

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> duration = end - start;

        std::cout << "C++ List Size: " << list_size << ", Time taken: " << duration.count() << " seconds" << std::endl;
    }

    return 0;
}

File: Tutorial-1/insertionSort/insertion_sort.rs
use std::time::Instant;

fn create_array(size: usize) -> Vec<usize> {
    let mut arr = Vec::with_capacity(size);
    for i in 0..size {
        arr.push(i * i);
    }
    arr
}

fn insertion_sort(arr: &mut [usize]) {
    let n = arr.len();
    for i in 1..n {
        let key = arr[i];
        let mut j = i;
        while j > 0 && arr[j - 1] > key {
            arr[j] = arr[j - 1];
            j -= 1;
        }
        arr[j] = key;
    }
}

fn main() {
    const START_SIZE: usize = 10000;
    const END_SIZE: usize = 100000;
    const STEP_SIZE: usize = 10000;

    for list_size in (START_SIZE..=END_SIZE).step_by(STEP_SIZE) {
        let mut numbers = create_array(list_size);

        let start = Instant::now();

        insertion_sort(&mut numbers);

        let duration = start.elapsed();

        println!("Rust List Size: {}, Time taken: {:.6} seconds", list_size, duration.as_secs_f64());
    }
}
